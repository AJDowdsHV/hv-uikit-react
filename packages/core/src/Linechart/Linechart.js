import React from "react";
import PropTypes from "prop-types";
import { withStyles } from "@material-ui/core";
import Chart from "../Chart";
import { setData, setLayout } from "./lineChartPlotlyOverrides";
import styles from "./styles";

const Linechart = ({
  id,
  classes,
  title,
  subtitle,
  data,
  layout,
  config,
  tooltipType = "multiple",
  type = "line",
  xAxisTitle,
  yAxisTitle,
  rangeSlider = false,
  ...others
}) => {
  const newData = setData(data, type);
  const newLayout = setLayout(layout, rangeSlider);

  return (
    <Chart
      id={id}
      classes={classes}
      title={title}
      subtitle={subtitle}
      xAxisTitle={xAxisTitle}
      yAxisTitle={yAxisTitle}
      data={newData}
      layout={newLayout}
      config={config}
      tooltipType={tooltipType}
      {...others}
    />
  );
};

Linechart.propTypes = {
  /**
   * An Id passed on to the component
   */
  id: PropTypes.string,
  /**
   * A Jss Object used to override or extend the styles applied.
   */
  classes: PropTypes.instanceOf(Object),
  /**
   * Title of the chart.
   */
  title: PropTypes.string,
  /**
   * Subtitle of the chart.
   */
  subtitle: PropTypes.string,
  /**
   * Plotly data object (see https://plot.ly/javascript/reference/).
   */
  data: PropTypes.arrayOf(PropTypes.instanceOf(Object)).isRequired,
  /**
   * Plotly layout object (see https://plot.ly/javascript/reference/#layout).
   */
  layout: PropTypes.instanceOf(Object),
  /**
   * Plotly config object (see https://plot.ly/javascript/configuration-options/).
   */
  config: PropTypes.instanceOf(Object),
  /**
   * Defines if should use a single or multiline tooltip.
   */
  tooltipType: PropTypes.oneOf(["single", "multiple"]),
  /**
   * Sets the type of graph.
   */
  type: PropTypes.oneOf(["line", "area", "stack"]),
  /**
   * Defines the title of the X axis.
   */
  xAxisTitle: PropTypes.string,
  /**
   * Defines the title of the Y axis.
   */
  yAxisTitle: PropTypes.string,
  /**
   * Defines it should present a range slider.
   */
  rangeSlider: PropTypes.bool
};

export default withStyles(styles, { name: "HvLinechart" })(Linechart);
