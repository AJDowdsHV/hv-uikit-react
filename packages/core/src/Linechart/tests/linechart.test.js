/* eslint-env jest */

import React from "react";
import { mount, shallow } from "enzyme";

import HvProvider from "../../Provider";
import { setData } from "../lineChartPlotlyOverrides";
import Linechart from "..";

describe("Linechart withStyles", () => {
  let wrapper;

  const trace1 = {
    x: [2300, 1000, 8500],
    y: ["January", "February", "March"],
    name: "Sales Target"
  };

  const data = [trace1];
  const layout = {};

  beforeEach(() => {
    wrapper = shallow(
      <HvProvider>
        <Linechart data={data} layout={layout} />
      </HvProvider>
    );
  });

  it("should be defined", () => {
    expect(wrapper).toBeDefined();
  });

  it("should render correctly", () => {
    expect(wrapper.find(Linechart)).toMatchSnapshot();
  });

  it("should render the Linechart", () => {
    wrapper = mount(
      <HvProvider>
        <Linechart data={data} layout={layout} />
      </HvProvider>
    );
    const linechart = wrapper.find(Linechart);
    expect(linechart.length).toBe(1);
  });

  it("should set default data for area", () => {
    const traceTest = {
      x: [2300, 1000, 8500],
      y: ["January", "February", "March"],
      name: "Sales Target"
    };

    const overwrittenData = setData([traceTest], "area");
    expect(overwrittenData[0].mode).toBe("lines");
    expect(overwrittenData[0].hoverinfo).toBe("none");
    expect(overwrittenData[0].fill).toBe("tonexty");
    expect(overwrittenData[0].type).toBe("scatter");
  });

  it("should set default data for line", () => {
    const traceTest = {
      x: [2300, 1000, 8500],
      y: ["January", "February", "March"],
      name: "Sales Target"
    };
    const overwrittenData = setData([traceTest], "line");
    expect(overwrittenData[0].mode).toBe("lines");
    expect(overwrittenData[0].hoverinfo).toBe("none");
    expect(overwrittenData[0].type).toBe("line");
  });

  it("should set default data for stack", () => {
    const traceTest = {
      x: [2300, 1000, 8500],
      y: ["January", "February", "March"],
      name: "Sales Target"
    };
    const overwrittenData = setData([traceTest], "stack");
    expect(overwrittenData[0].mode).toBe("lines");
    expect(overwrittenData[0].hoverinfo).toBe("none");
    expect(overwrittenData[0].fill).toBe("tonexty");
    expect(overwrittenData[0].type).toBe("scatter");
    expect(overwrittenData[0].stackgroup).toBe("one");
  });

  it("shouldn't override existing data", () => {
    const traceTest = { hoverinfo: "xx" };
    const overwrittenData = setData([traceTest]);
    expect(overwrittenData[0].hoverinfo).toBe("xx");
  });
});
