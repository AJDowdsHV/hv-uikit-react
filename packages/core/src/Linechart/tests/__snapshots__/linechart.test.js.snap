// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Linechart should render correctly 1`] = `
<WithStyles(Linechart)
  data={
    Array [
      Object {
        "hoverinfo": "none",
        "mode": "lines",
        "name": "Sales Target",
        "type": "line",
        "x": Array [
          "January",
          "February",
          "March",
          "April",
          "May",
          "June",
          "July",
          "August",
          "September",
          "October",
          "November",
          "December",
        ],
        "y": Array [
          5929,
          2393,
          1590,
          7817,
          4749,
          1702,
          2381,
          2909,
          6732,
          3098,
          2119,
          2146,
        ],
      },
    ]
  }
  subtitle="Sales performance (YTD)"
  title="Simple line"
  xAxisTitle="2018"
  yAxisTitle="Thousands of Dollars ($)"
>
  <Linechart
    classes={Object {}}
    data={
      Array [
        Object {
          "hoverinfo": "none",
          "mode": "lines",
          "name": "Sales Target",
          "type": "line",
          "x": Array [
            "January",
            "February",
            "March",
            "April",
            "May",
            "June",
            "July",
            "August",
            "September",
            "October",
            "November",
            "December",
          ],
          "y": Array [
            5929,
            2393,
            1590,
            7817,
            4749,
            1702,
            2381,
            2909,
            6732,
            3098,
            2119,
            2146,
          ],
        },
      ]
    }
    subtitle="Sales performance (YTD)"
    title="Simple line"
    xAxisTitle="2018"
    yAxisTitle="Thousands of Dollars ($)"
  >
    <WithStyles(Chart)
      classes={Object {}}
      data={
        Array [
          Object {
            "hoverinfo": "none",
            "mode": "lines",
            "name": "Sales Target",
            "type": "line",
            "x": Array [
              "January",
              "February",
              "March",
              "April",
              "May",
              "June",
              "July",
              "August",
              "September",
              "October",
              "November",
              "December",
            ],
            "y": Array [
              5929,
              2393,
              1590,
              7817,
              4749,
              1702,
              2381,
              2909,
              6732,
              3098,
              2119,
              2146,
            ],
          },
        ]
      }
      layout={Object {}}
      subtitle="Sales performance (YTD)"
      title="Simple line"
      tooltipType="multiple"
      xAxisTitle="2018"
      yAxisTitle="Thousands of Dollars ($)"
    >
      <Chart
        classes={
          Object {
            "defaultColors": "HvChart-defaultColors",
            "gridColor": "HvChart-gridColor",
            "lineColor": "HvChart-lineColor",
            "paddingTop": "HvChart-paddingTop",
            "plotColor": "HvChart-plotColor",
            "root": "HvChart-root",
            "subtitle": "HvChart-subtitle",
            "titleContainer": "HvChart-titleContainer",
            "vizText": "HvChart-vizText",
          }
        }
        data={
          Array [
            Object {
              "hoverinfo": "none",
              "mode": "lines",
              "name": "Sales Target",
              "type": "line",
              "x": Array [
                "January",
                "February",
                "March",
                "April",
                "May",
                "June",
                "July",
                "August",
                "September",
                "October",
                "November",
                "December",
              ],
              "y": Array [
                5929,
                2393,
                1590,
                7817,
                4749,
                1702,
                2381,
                2909,
                6732,
                3098,
                2119,
                2146,
              ],
            },
          ]
        }
        layout={Object {}}
        subtitle="Sales performance (YTD)"
        title="Simple line"
        tooltipType="multiple"
        xAxisTitle="2018"
        yAxisTitle="Thousands of Dollars ($)"
      >
        <div
          className="HvChart-root"
        >
          <div
            className="HvChart-titleContainer"
          >
            <WithStyles(ForwardRef)
              variant="mTitle"
            >
              <ForwardRef
                classes={
                  Object {
                    "3xlTitle": "HvTypography-3xlTitle",
                    "4xlTitle": "HvTypography-4xlTitle",
                    "5xlTitle": "HvTypography-5xlTitle",
                    "disabledText": "HvTypography-disabledText",
                    "highlightText": "HvTypography-highlightText",
                    "lTitle": "HvTypography-lTitle",
                    "link": "HvTypography-link",
                    "mTitle": "HvTypography-mTitle",
                    "normalText": "HvTypography-normalText",
                    "placeholderText": "HvTypography-placeholderText",
                    "root": "HvTypography-root",
                    "sTitle": "HvTypography-sTitle",
                    "selectedNavText": "HvTypography-selectedNavText",
                    "vizText": "HvTypography-vizText",
                    "vizTextDisabled": "HvTypography-vizTextDisabled",
                    "xlTitle": "HvTypography-xlTitle",
                    "xsInlineLink": "HvTypography-xsInlineLink",
                    "xsTitle": "HvTypography-xsTitle",
                    "xxlTitle": "HvTypography-xxlTitle",
                    "xxsTitle": "HvTypography-xxsTitle",
                  }
                }
                variant="mTitle"
              >
                <h3
                  className="HvTypography-root HvTypography-mTitle"
                >
                  Simple line
                </h3>
              </ForwardRef>
            </WithStyles(ForwardRef)>
            <div
              className="HvChart-subtitle"
            >
              <WithStyles(ForwardRef)
                variant="sText"
              >
                <ForwardRef
                  classes={
                    Object {
                      "3xlTitle": "HvTypography-3xlTitle",
                      "4xlTitle": "HvTypography-4xlTitle",
                      "5xlTitle": "HvTypography-5xlTitle",
                      "disabledText": "HvTypography-disabledText",
                      "highlightText": "HvTypography-highlightText",
                      "lTitle": "HvTypography-lTitle",
                      "link": "HvTypography-link",
                      "mTitle": "HvTypography-mTitle",
                      "normalText": "HvTypography-normalText",
                      "placeholderText": "HvTypography-placeholderText",
                      "root": "HvTypography-root",
                      "sTitle": "HvTypography-sTitle",
                      "selectedNavText": "HvTypography-selectedNavText",
                      "vizText": "HvTypography-vizText",
                      "vizTextDisabled": "HvTypography-vizTextDisabled",
                      "xlTitle": "HvTypography-xlTitle",
                      "xsInlineLink": "HvTypography-xsInlineLink",
                      "xsTitle": "HvTypography-xsTitle",
                      "xxlTitle": "HvTypography-xxlTitle",
                      "xxsTitle": "HvTypography-xxsTitle",
                    }
                  }
                  variant="sText"
                >
                  <span
                    className="HvTypography-root"
                  >
                    Sales performance (YTD)
                  </span>
                </ForwardRef>
              </WithStyles(ForwardRef)>
            </div>
          </div>
          <div
            className="HvChart-paddingTop"
            onMouseMove={[Function]}
          >
            <WithStyles(Plot)
              config={
                Object {
                  "displayModeBar": false,
                  "responsive": true,
                }
              }
              data={
                Array [
                  Object {
                    "hoverinfo": "none",
                    "mode": "lines",
                    "name": "Sales Target",
                    "type": "line",
                    "x": Array [
                      "January",
                      "February",
                      "March",
                      "April",
                      "May",
                      "June",
                      "July",
                      "August",
                      "September",
                      "October",
                      "November",
                      "December",
                    ],
                    "y": Array [
                      5929,
                      2393,
                      1590,
                      7817,
                      4749,
                      1702,
                      2381,
                      2909,
                      6732,
                      3098,
                      2119,
                      2146,
                    ],
                  },
                ]
              }
              layout={
                Object {
                  "autosize": true,
                  "colorway": Array [
                    "#6EAFFF",
                    "#FFAB5C",
                    "#5CD1B2",
                    "#9672C1",
                    "#ED6868",
                    "#58C9DD",
                    "#FFDB70",
                    "#4AB573",
                    "#646A98",
                    "#EB7397",
                    "#33ABCC",
                    "#F8C169",
                    "#DE64BD",
                    "#419393",
                    "#F5996E",
                    "#528BBF",
                    "#EB95AF",
                    "#73BF9C",
                    "#B55984",
                    "#F7B552",
                  ],
                  "legend": Object {
                    "font": Object {
                      "color": "#414141",
                      "family": Array [
                        "'Open Sans'",
                        "Arial",
                        "Helvetica",
                        "sans-serif",
                      ],
                      "size": 10,
                    },
                    "orientation": "h",
                    "x": 1,
                    "xanchor": "right",
                    "y": 1.1,
                    "yanchor": "bottom",
                  },
                  "margin": Object {
                    "b": 50,
                    "l": 50,
                    "pad": 0,
                    "t": 10,
                  },
                  "paper_bgcolor": "#FBFCFC",
                  "plot_bgcolor": "#FBFCFC",
                  "xaxis": Object {
                    "automargin": true,
                    "autorange": true,
                    "fixedrange": true,
                    "gridcolor": "#999999",
                    "linecolor": "#CCCED0",
                    "linewidth": 1,
                    "range": Array [
                      0,
                      11,
                    ],
                    "showgrid": false,
                    "tickcolor": "#CCCED0",
                    "tickfont": Object {
                      "color": "#414141",
                      "family": Array [
                        "'Open Sans'",
                        "Arial",
                        "Helvetica",
                        "sans-serif",
                      ],
                      "size": 10,
                    },
                    "ticklen": 2,
                    "title": Object {
                      "font": Object {
                        "color": "#414141",
                        "family": Array [
                          "'Open Sans'",
                          "Arial",
                          "Helvetica",
                          "sans-serif",
                        ],
                        "size": 10,
                      },
                      "text": "2018",
                    },
                    "type": "category",
                    "zeroline": false,
                  },
                  "yaxis": Object {
                    "automargin": true,
                    "autorange": true,
                    "fixedrange": true,
                    "gridcolor": "#999999",
                    "linecolor": "#CCCED0",
                    "linewidth": 1,
                    "range": Array [
                      1244.0555555555557,
                      8162.944444444444,
                    ],
                    "showgrid": true,
                    "tickcolor": "#CCCED0",
                    "tickfont": Object {
                      "color": "#414141",
                      "family": Array [
                        "'Open Sans'",
                        "Arial",
                        "Helvetica",
                        "sans-serif",
                      ],
                      "size": 10,
                    },
                    "ticklen": 2,
                    "title": Object {
                      "font": Object {
                        "color": "#414141",
                        "family": Array [
                          "'Open Sans'",
                          "Arial",
                          "Helvetica",
                          "sans-serif",
                        ],
                        "size": 10,
                      },
                      "text": "Thousands of Dollars ($)",
                    },
                    "type": "linear",
                    "zeroline": false,
                  },
                }
              }
              onHover={[Function]}
              onUnHover={[Function]}
              title="Simple line"
            >
              <Plot
                classes={Object {}}
                config={
                  Object {
                    "displayModeBar": false,
                    "responsive": true,
                  }
                }
                data={
                  Array [
                    Object {
                      "hoverinfo": "none",
                      "mode": "lines",
                      "name": "Sales Target",
                      "type": "line",
                      "x": Array [
                        "January",
                        "February",
                        "March",
                        "April",
                        "May",
                        "June",
                        "July",
                        "August",
                        "September",
                        "October",
                        "November",
                        "December",
                      ],
                      "y": Array [
                        5929,
                        2393,
                        1590,
                        7817,
                        4749,
                        1702,
                        2381,
                        2909,
                        6732,
                        3098,
                        2119,
                        2146,
                      ],
                    },
                  ]
                }
                layout={
                  Object {
                    "autosize": true,
                    "colorway": Array [
                      "#6EAFFF",
                      "#FFAB5C",
                      "#5CD1B2",
                      "#9672C1",
                      "#ED6868",
                      "#58C9DD",
                      "#FFDB70",
                      "#4AB573",
                      "#646A98",
                      "#EB7397",
                      "#33ABCC",
                      "#F8C169",
                      "#DE64BD",
                      "#419393",
                      "#F5996E",
                      "#528BBF",
                      "#EB95AF",
                      "#73BF9C",
                      "#B55984",
                      "#F7B552",
                    ],
                    "legend": Object {
                      "font": Object {
                        "color": "#414141",
                        "family": Array [
                          "'Open Sans'",
                          "Arial",
                          "Helvetica",
                          "sans-serif",
                        ],
                        "size": 10,
                      },
                      "orientation": "h",
                      "x": 1,
                      "xanchor": "right",
                      "y": 1.1,
                      "yanchor": "bottom",
                    },
                    "margin": Object {
                      "b": 50,
                      "l": 50,
                      "pad": 0,
                      "t": 10,
                    },
                    "paper_bgcolor": "#FBFCFC",
                    "plot_bgcolor": "#FBFCFC",
                    "xaxis": Object {
                      "automargin": true,
                      "autorange": true,
                      "fixedrange": true,
                      "gridcolor": "#999999",
                      "linecolor": "#CCCED0",
                      "linewidth": 1,
                      "range": Array [
                        0,
                        11,
                      ],
                      "showgrid": false,
                      "tickcolor": "#CCCED0",
                      "tickfont": Object {
                        "color": "#414141",
                        "family": Array [
                          "'Open Sans'",
                          "Arial",
                          "Helvetica",
                          "sans-serif",
                        ],
                        "size": 10,
                      },
                      "ticklen": 2,
                      "title": Object {
                        "font": Object {
                          "color": "#414141",
                          "family": Array [
                            "'Open Sans'",
                            "Arial",
                            "Helvetica",
                            "sans-serif",
                          ],
                          "size": 10,
                        },
                        "text": "2018",
                      },
                      "type": "category",
                      "zeroline": false,
                    },
                    "yaxis": Object {
                      "automargin": true,
                      "autorange": true,
                      "fixedrange": true,
                      "gridcolor": "#999999",
                      "linecolor": "#CCCED0",
                      "linewidth": 1,
                      "range": Array [
                        1244.0555555555557,
                        8162.944444444444,
                      ],
                      "showgrid": true,
                      "tickcolor": "#CCCED0",
                      "tickfont": Object {
                        "color": "#414141",
                        "family": Array [
                          "'Open Sans'",
                          "Arial",
                          "Helvetica",
                          "sans-serif",
                        ],
                        "size": 10,
                      },
                      "ticklen": 2,
                      "title": Object {
                        "font": Object {
                          "color": "#414141",
                          "family": Array [
                            "'Open Sans'",
                            "Arial",
                            "Helvetica",
                            "sans-serif",
                          ],
                          "size": 10,
                        },
                        "text": "Thousands of Dollars ($)",
                      },
                      "type": "linear",
                      "zeroline": false,
                    },
                  }
                }
                onHover={[Function]}
                onUnHover={[Function]}
                title="Simple line"
              >
                <PlotlyComponent
                  classes={Object {}}
                  config={
                    Object {
                      "displayModeBar": false,
                      "responsive": true,
                    }
                  }
                  data={
                    Array [
                      Object {
                        "hoverinfo": "none",
                        "mode": "lines",
                        "name": "Sales Target",
                        "type": "line",
                        "x": Array [
                          "January",
                          "February",
                          "March",
                          "April",
                          "May",
                          "June",
                          "July",
                          "August",
                          "September",
                          "October",
                          "November",
                          "December",
                        ],
                        "y": Array [
                          5929,
                          2393,
                          1590,
                          7817,
                          4749,
                          1702,
                          2381,
                          2909,
                          6732,
                          3098,
                          2119,
                          2146,
                        ],
                      },
                    ]
                  }
                  debug={false}
                  layout={
                    Object {
                      "autosize": true,
                      "colorway": Array [
                        "#6EAFFF",
                        "#FFAB5C",
                        "#5CD1B2",
                        "#9672C1",
                        "#ED6868",
                        "#58C9DD",
                        "#FFDB70",
                        "#4AB573",
                        "#646A98",
                        "#EB7397",
                        "#33ABCC",
                        "#F8C169",
                        "#DE64BD",
                        "#419393",
                        "#F5996E",
                        "#528BBF",
                        "#EB95AF",
                        "#73BF9C",
                        "#B55984",
                        "#F7B552",
                      ],
                      "legend": Object {
                        "font": Object {
                          "color": "#414141",
                          "family": Array [
                            "'Open Sans'",
                            "Arial",
                            "Helvetica",
                            "sans-serif",
                          ],
                          "size": 10,
                        },
                        "orientation": "h",
                        "x": 1,
                        "xanchor": "right",
                        "y": 1.1,
                        "yanchor": "bottom",
                      },
                      "margin": Object {
                        "b": 50,
                        "l": 50,
                        "pad": 0,
                        "t": 10,
                      },
                      "paper_bgcolor": "#FBFCFC",
                      "plot_bgcolor": "#FBFCFC",
                      "xaxis": Object {
                        "automargin": true,
                        "autorange": true,
                        "fixedrange": true,
                        "gridcolor": "#999999",
                        "linecolor": "#CCCED0",
                        "linewidth": 1,
                        "range": Array [
                          0,
                          11,
                        ],
                        "showgrid": false,
                        "tickcolor": "#CCCED0",
                        "tickfont": Object {
                          "color": "#414141",
                          "family": Array [
                            "'Open Sans'",
                            "Arial",
                            "Helvetica",
                            "sans-serif",
                          ],
                          "size": 10,
                        },
                        "ticklen": 2,
                        "title": Object {
                          "font": Object {
                            "color": "#414141",
                            "family": Array [
                              "'Open Sans'",
                              "Arial",
                              "Helvetica",
                              "sans-serif",
                            ],
                            "size": 10,
                          },
                          "text": "2018",
                        },
                        "type": "category",
                        "zeroline": false,
                      },
                      "yaxis": Object {
                        "automargin": true,
                        "autorange": true,
                        "fixedrange": true,
                        "gridcolor": "#999999",
                        "linecolor": "#CCCED0",
                        "linewidth": 1,
                        "range": Array [
                          1244.0555555555557,
                          8162.944444444444,
                        ],
                        "showgrid": true,
                        "tickcolor": "#CCCED0",
                        "tickfont": Object {
                          "color": "#414141",
                          "family": Array [
                            "'Open Sans'",
                            "Arial",
                            "Helvetica",
                            "sans-serif",
                          ],
                          "size": 10,
                        },
                        "ticklen": 2,
                        "title": Object {
                          "font": Object {
                            "color": "#414141",
                            "family": Array [
                              "'Open Sans'",
                              "Arial",
                              "Helvetica",
                              "sans-serif",
                            ],
                            "size": 10,
                          },
                          "text": "Thousands of Dollars ($)",
                        },
                        "type": "linear",
                        "zeroline": false,
                      },
                    }
                  }
                  onAfterPlot={null}
                  onHover={[Function]}
                  onUnhover={[Function]}
                  revision={0}
                  style={
                    Object {
                      "position": "relative",
                    }
                  }
                  title="Simple line"
                  useResizeHandler={false}
                >
                  <div
                    style={
                      Object {
                        "position": "relative",
                      }
                    }
                  />
                </PlotlyComponent>
              </Plot>
            </WithStyles(Plot)>
          </div>
        </div>
      </Chart>
    </WithStyles(Chart)>
  </Linechart>
</WithStyles(Linechart)>
`;
